### ISSUE TITLE
Inbound packets require a linear scan of all ports to deliver packets

### ISSUE THREAD

**Issue by FelixMcFelix** [oxidecomputer/opte#637]
In the packet receive path, we perform a linear scan over all ports to find a matching VNI + destination MAC address for every packet.

https://github.com/oxidecomputer/opte/blob/9b57c4bbb761cd1d347b9a912ae865324ffe3c78/xde/src/xde.rs#L1445-L1447

This does have a significant performance impact for the last port in the list, as measured on `glasgow`:

| $n$ additional ports | iPerf3 throughput |
|--------|--------|
| 0 | ~3.00 Gbps |
| 32 | ~2.82 Gbps |
| 64 | ~2.72 Gbps |
| 256 | ~2.36 Gbps |
| 512 | ~1.90 Gbps |
| 1024 | ~1.22 Gbps |

While 256 ports is maybe a bit on the extreme end today, port counts in the range 64–128 are not unreasonable. In theory this number can grow fairly large – one for each service, ≥1 for each instance (max 128 instances today, I believe). On higher density Turin parts we're looking at more threads (384+), so more ports is likely when O(instances) ≈ O(threads).

We should use a `BTreeMap`, presort the `Vec<Box<XdeDev>>` on `(Vni, MacAddr)`, or keep an adjacent lookup table. [`slice::binary_search_by_key`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search_by_key) will be key for the latter – which is best will require some benchmarking, since we have varying levels of indirection here. We could do the same for port names (add/removal requires a linear scan here too), but this is not a hotpath.

**Comment by FelixMcFelix** [comment]
I threw together [something quick](https://gist.github.com/FelixMcFelix/81af609812df9c5527395a6dddf937ea) to test approximate performance using similarly sized types and the same levels of indirection. There are a few axes here:
* Sorted `Vec` vs `BTreeMap`
* `(Vni, MacAddr)` (A) vs `(u32, u64)` (B) as key,
* Storing `idx` to unsorted/name-sorted vec.

| $n_{ports}$ | Linear(best)/A |  Linear(best)/B | Linear(worst)/A | Linear(worst)/B | `BTreeMap`/A | `BTreeMap`/B | `BTreeMap`/B/`idx` | `Vec`/A | `Vec`/B | `Vec`/A/`idx` | `Vec`/B/`idx` |
|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|
| 1 | 1.5360 | 1.1502 | 1.5114 | 1.1027 | 2.5919 | 1.4148 | 1.9443 | 7.1392 | 6.6873 | 2.1840 | 1.2094 |
| 4 | – | – | 3.0061 | 2.8403 | 4.3731 | 2.9436 | 3.7066 | 9.9132 | 8.3890 | 6.4299 | 3.3996 |
| 8 | – | – | 5.0862 | 5.2861 | 7.1841 | 5.2621 | 6.3939 | 14.600 | 10.398 | 9.4509 | 5.3492 |
| 16 | – | – | 8.4726 | 9.4915 | 5.8975 | 4.4617 | 5.0088 | 20.167 | 13.094 | 11.911 | 7.0289 |
| 32 | – | – | 16.076 | 19.695 | 10.013 | 6.7277 | 7.6258 | 22.413 | 18.979 | 14.875 | 9.4967 |
| 64 | – | – | 32.173 | 40.677 | 8.2921 | 5.8758 | 6.9861 | 30.112 | 21.734 | 20.487 | 11.360 |
| 128 | – | – | 65.920 | 77.841 | 14.441 | 10.427 | 11.901 | 39.291 | 26.814 | 24.033 | 13.368 |
| 256 | – | – | 140.32 | 162.39 | 22.947 | 14.957 | 16.626 | 44.097 | 31.891 | 27.837 | 15.558 |
| 1024 | – | – | 817.96 | 929.26 | 31.268 | 19.465 | 21.812 | 52.281 | 41.672 | 35.947 | 20.938 |

All times in ns, run on my helios x86 workstation (AMD Ryzen 7 5700G), though I basically saw the same trends on my M2 Macbook Pro. I'm using a seeded mac+VNI pool, so we shouldn't have inconsistency there (same needle + haystack per search).

To summarise, it looks like `BTreeMap`/B wins outright -- `BTreeMap` using `(u32, u64)` as the key. I assume this is down to far faster `Ord`/`Eq` implementations. I thought one of the other approaches might have pulled ahead for opaque cache adjacency reasons, but it's good to have concrete data on this one.
